src/pulsar_client_worker.erl

```diff
@@ -0,0 +1,576 @@
+    case get_alive_sock_opts(Server, undefined, Opts) of
+        {error, Reason} ->
+            %% to avoid a hot restart loop leading to max restart
+            %% intensity when pulsar (or the connection to it) is
+            %% having a bad day...
+            Parent ! {self(), {error, Reason}},
+            timer:sleep(100),
+            ?MODULE:try_initial_connection(Parent, Server, Opts);
+        {ok, {Sock, Opts1}} ->
+            pulsar_socket:controlling_process(Sock, Parent, Opts1),
+            Parent ! {self(), {ok, {Sock, Opts1}}}
+    end.
+
+get_service_url_from_lookup_response(Response, Opts) ->
+    case {Opts, Response} of
+        {#{enable_ssl := true}, #{brokerServiceUrlTls := BrokerServiceUrlTls}} ->
+            BrokerServiceUrlTls;
+        {#{enable_ssl := true}, #{brokerServiceUrl := BrokerServiceUrl}} ->
+            log_error("SSL enabled but brokerServiceUrlTls is not provided by pulsar,"
+                      " falling back to brokerServiceUrl: ~p", [BrokerServiceUrl]),
+            BrokerServiceUrl;
+        {_, #{brokerServiceUrl := BrokerServiceUrl}} ->
+            %% the 'brokerServiceUrl' is a mandatory field in case the SSL is disabled
+            BrokerServiceUrl
+    end.
+
+%% If we receive a response of lookup type `Redirect', we must re-issue the lookup
+%% connecting to the returned broker.
+%% https://pulsar.apache.org/docs/2.11.x/developing-binary-protocol/#lookuptopicresponse
+handle_redirect_lookup_response(State, From, Topic, Response) ->
+    #state{client_id = ClientId, opts = Opts} = State,
+    ServiceURL = get_service_url_from_lookup_response(Response, Opts),
+    {ok, Sup} = pulsar_client_sup:find_worker_sup(ClientId),
+    case pulsar_client_worker_sup:start_worker(Sup, ClientId, ServiceURL, Opts) of
+        {error, Reason} ->
+            log_error("failed to ensure worker for new url is started;"
+                      " reason: ~p ; client_id: ~p ; new url: ~s ; topic: ~p",
+                      [Reason, ClientId, ServiceURL, Topic]),
+            gen_server:reply(From, {error, {failed_to_start_worker, Reason}}),
+            State;
+        {ok, WorkerPid} ->
+            lookup_topic_redirect_async(WorkerPid, Topic, From, Response),
+            State
+    end.
+
+lookup_topic_redirect_async(WorkerPid, PartitionTopic, OriginalFrom, Response) ->
+    Opts = #{authoritative => maps:get(authoritative, Response, false)},
+    gen_server:cast(WorkerPid, #lookup_topic_async{
+                                  from = OriginalFrom,
+                                  partition_topic = PartitionTopic,
+                                  opts = Opts
+                                 }),
+    ok.
+
+connection_succeeded(State0) ->
+    State0#state{connection_failure_count = 0}.
+
+handle_connection_failure(State0) ->
+    Count = State0#state.connection_failure_count,
+    case Count > ?MAX_CONNECTION_FAILURES of
+        true ->
+            stop;
```

Comment:
Should we only stop if this was dynamically spawned, and not one of the "bootstrap server" clients?